rfcs/0164-nim-library-lockfiles.md

```diff
@@ -0,0 +1,142 @@
+It also shifts the labor required to package Nim programs from maintaining a collection of FOD expressions to maintaining tooling that can automatically reason over what dependencies are required for a given program.
+
+# Detailed design
+[design]: #detailed-design
+
+- Implement and maintain a tool for generating Nix specific lockfiles using the standard packaging metadata local to Nim programs. At the moment this is the [nim_lk](https://git.sr.ht/~ehmry/nim_lk) tool.
+
+- Commit lockfiles into Nixpkgs.
+
+- Expand lockfiles at evaluation time to collections of fetch expressions using `fetchzip`, `fetchgit`, `fetchhg`, etc.
+
+- Augment libraries selected by lockfiles so that Nim wrapper libraries include additional metadata such as `buildInputs` to include shared libraries from non-Nim packages.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Old way:
+```nix
+# <nixpkgs>/pkgs/tools/system/ttop/default.nix
+{ lib, nimPackages, fetchFromGitHub }:
+
+nimPackages.buildNimPackage (finalAttrs: {
+  pname = "ttop";
+  version = "1.0.1";
+  nimBinOnly = true;
+
+  src = fetchFromGitHub { /* … */ };
+
+  buildInputs = with nimPackages; [ asciigraph illwill parsetoml zippy ];
+})
+```
+
+## New way:
+```nix
+# <nixpkgs>/pkgs/by-name/tt/ttop/package.nix
+{ lib, buildNimPackage, fetchFromGitHub }:
+
+buildNimPackage (finalAttrs: {
+  pname = "ttop";
+  version = "1.0.1";
+
+  src = fetchFromGitHub { /* … */ };
+
+  buildNimLockfile = ./lock.json;
+    # lock.json must be checked into nixpkgs at
+    # <nixpkgs>/pkgs/by-name/tt/ttop/lock.json
+})
+```
+
+## Example lockfile:
+```json
+{
+  "depends": [
+    {
+      "method": "fetchurl",
+      "path": "/nix/store/70cqa9s36dqnmsf179cn9psj77jhqi1l-source",
+      "rev": "a4a1affd45ba90bea24e08733ae2bd02fe058166",
+      "sha256": "005ib6im97x9pdbg6p0fy58zpdwdbkpmilxa8nhrrb1hnpjzz90p",
+      "url": "https://git.sr.ht/~ehmry/nim_cbor/archive/a4a1affd45ba90bea24e08733ae2bd02fe058166.tar.gz",
+      "ref": "20230619",
+      "packages": [
+        "cbor"
+      ],
+      "srcDir": "src"
+    },
+    {
+      "method": "fetchurl",
+      "path": "/nix/store/ffkxmjmigfs7zhhiiqm0iw2c34smyciy-source",
+      "rev": "26d62fdc40feb84c6533956dc11d5ee9ea9b6c09",
+      "sha256": "0xpzifjkfp49w76qmaylan8q181bs45anmp46l4bwr3lkrr7bpwh",
+      "url": "https://github.com/zevv/npeg/archive/26d62fdc40feb84c6533956dc11d5ee9ea9b6c09.tar.gz",
+      "ref": "1.2.1",
+      "packages": [
+        "npeg"
+      ],
+      "srcDir": "src"
+    }
+  ]
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Lockfiles must be checked into Nixpkgs which would bloat the repository.
+- Checks on Nim libraries would not be run (unless implemented in `buildNimPackage`).
+- Nim programs would be updated as usual but their dependencies would not. Updating lockfiles would be non-trivial for automated tools such as `r-ryantm/nixpkgs-update`.
```

Comment:
This could be solved by making `buildNimPackage` set `passthru.updateScript` to a sane default